### DISCUSSION TITLE
Avoiding malloc when interfacing with json and asyncio sockets

### DISCUSSION THREAD

**Discussion by mzakharocsc** [micropython/micropython#15527] [Ideas]
When faced with low-memory constrained systems, and sending large json payloads over async MQTT sockets, it is essential to avoid malloc of large buffers along the transmit path. Unfortunately, micropython currently does not provide malloc-free API to accomplish such goals. 

First problem is  converting a python datastructure to json bytes and reusing an existing buffer.  One could accomplish such goal with the following API:
```
buf = io.BytesIO(maxlen)
def get_payload(data):
  buf.seek(0)
  buf.truncate(0)
  json.dump(data, buf)
  payload = buf.getvalue()
 ```

Unfortunately,  `buf.truncate(0)` API does not exist in micropython, nor will  `buf.getvalue()` be malloc-free. 

Proposal 1: add `.truncate(val)` API to io.BytesIO class in micropython, to allow resetting the buffer to 0 every time it is used

Proposal 2: create a variant of `buf.getvalue(None)` with optional parameter that will be a micropython extension to return payload without additional malloc - pointer to the data.

Finally, sending the payload over the socket is hamstrung  by asyncio `write()` and `drain()` API. asyncio  `write()` function builds an internal [`self.out_buf`](https://github.com/micropython/micropython/blob/master/extmod/asyncio/stream.py#L74) buffer which also calls malloc.  

Proposal 3: add a  asyncio `write_and_drain()` API to write to socket in one go and avoid extra malloc.

We have implemented local versions of the 3 proposals in our attempt to build a json-async-mqtt-socket interface that reuses a single buffer for communicating large payloads. It allowed us to successfully target systems with very low available memory, which was not possible before. Question is whether proposed changes would be useful to the community at large.

**Comment by peterhinch** [comment]
Re asyncio see https://github.com/micropython/micropython/pull/7868 - in particular the response of @dpgeorge.

**Reply by mzakharocsc** [reply]
looks like https://github.com/micropython/micropython/pull/10837 tries to fix asyncio issue.
